ts/Base/MaterialComponent.ts

```diff
@@ -0,0 +1,134 @@
+import {MDCRipple} from '@material/ripple';
+import autobind from 'autobind-decorator';
+import {Component, VNode} from 'preact';
+
+export interface IMaterialComponentProps extends JSX.HTMLAttributes {
+  ripple?: boolean;
+}
+
+export interface IMaterialComponentState {}
+
+/**
+ * Base class for every Material component in this package
+ * NOTE: every component should add a ref by the name of `control` to its root dom for autoInit Properties
+ *
+ * @export
+ * @class MaterialComponent
+ * @extends {Component}
+ */
+export abstract class MaterialComponent<
+  PropsType extends {[prop: string]: any},
+  StateType extends {[prop: string]: any}
+> extends Component<
+  PropsType & IMaterialComponentProps,
+  StateType & IMaterialComponentState
+> {
+  /**
+   * Attributes inside this array will be check for boolean value true
+   * and will be converted to mdc classes
+   */
+  protected abstract mdcProps: string[];
+  /** This will again be used to add apt classname to the component */
+  protected abstract componentName: string;
+
+  /** The final class name given to the dom */
+  protected classText: string;
+
+  // TODO: find out correct type
+  protected control?: any;
+
+  public render(props: PropsType & IMaterialComponentProps): VNode {
+    if (!this.classText) {
+      this.classText = this.buildClassName();
+    }
+    // Fetch a VNode
+    const componentProps = props;
+    const userDefinedClasses =
+      componentProps.className || componentProps.class || '';
+
+    // We delete class props and add them later in the final
+    // step so every component does not need to handle user specified classes.
+    if (componentProps.class) {
+      delete componentProps.class;
+    }
+    if (componentProps.className) {
+      delete componentProps.className;
+    }
+
+    const element = this.materialDom(componentProps);
+    element.attributes = element.attributes || {};
+
+    element.attributes.className = `${userDefinedClasses} ${this.getClassName(
+      element
+    )}`
+      .split(' ')
+      .filter(
+        (value, index, self) => self.indexOf(value) === index && value !== ''
+      ) // Unique + exclude empty class names
+      .join(' ');
+    // Clean this shit of proxy attributes
+    this.mdcProps.forEach(prop => {
+      delete element.attributes[prop];
+    });
+    return element;
+  }
+
+  /** Attach the ripple effect */
+  @autobind
+  protected attachRipple() {
+    if (this.props.ripple && this.control) {
+      MDCRipple.attachTo(this.control);
```

Comment:
Just looked more into what the `MDCRipple.attachTo()` does and it seems like it's adding event listeners (https://github.com/material-components/material-components-web/blob/master/packages/mdc-ripple/foundation.js).

This means that we really can't use `attachTo()` as it would introduce memory leaks if an object is removed as the ripple is not. I would recommend that we instead construct the MDCRipple ourselves and store the instance. We would then need to call `destroy()` on it on `componentWillUnmount()`. You will also be required to call `super.componentWillUnmount()` in every `componentWillUnmount()` method.

Reply:
Can't attach ripple be componentDidMount?

Reply:
Yes, I would think so.